baremetal-coco: tdx: Add DCAP deployment

fidencio opened the pull request:
Please, see each commit message for a better explanation of what's being done.

This is out for early testing, and there are a few things to consider:
* We will **NOT** require any SELinux changes, as I was able to work that requirement around with an init container
* Tests must be performed on multi-node setup, as so far it's only been tested on single node ones

I will give instructions to @bpradipt, and give him access to a clean SNO cluster, so he can go ahead and give it a try during the time I will be off.

**KNOWN ISSUES**:
- [x] Registration on a factory reset machine leads to the following error:
  * From the registration pods logs:
  ```
  Error: unexpected error occurred while sending data to cache server.
  Error: the data couldn't be sent to cache server!

  Intel(R) Software Guard Extensions PCK Cert ID Retrieval Tool Version 1.22.100.3

  Warning: platform manifest is not available or current platform is not multi-package platform.
  ```
  * From the pccs pod logs:
  ```
  2024-12-19 19:03:00.253 [info]: Client Request-ID : 27fdde775ad94d3e8274ed805ab4ba8b
  2024-12-19 19:03:00.257 [info]: Executing (default): SELECT `qe_id`, `pce_id`, `platform_manifest`, `enc_ppid`, `fmspc`, `ca`, `created_time`, `updated_time` FROM `platforms` AS `Platforms` WHERE `Platforms`.`qe_id` = '16D27A5828C08A20BAE67D9EA535B88B' AND `Platforms`.`pce_id` = '0000';
  2024-12-19 19:03:00.582 [info]: Request-ID is : 4b80d2c7d79b4357bda779a72422dc89
  2024-12-19 19:03:00.582 [debug]: Request URL https://api.trustedservices.intel.com/sgx/certification/v4/pckcerts
  2024-12-19 19:03:00.583 [error]: Intel PCS server returns error(404).
  2024-12-19 19:03:00.583 [error]: Error: No cache data for this platform.
    at Module.getPckCertFromPCS (file:///opt/intel/pccs/services/logic/commonCacheLogic.js:159:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async LazyCachingMode.registerPlatforms (file:///opt/intel/pccs/services/caching_modes/cachingMode.js:163:7)
    at async Module.registerPlatforms (file:///opt/intel/pccs/services/platformsRegService.js:107:3)
    at async postPlatforms (file:///opt/intel/pccs/controllers/platformsController.js:52:5)
  2024-12-19 19:03:00.586 [info]: 10.128.0.40 - - [19/Dec/2024:19:03:00 +0000] "POST /sgx/certification/v4/platforms HTTP/1.1" 404 32 "-" "-"
  ```
- [x] registration pod may start before the pccs pod after a reboot

fidencio committed:
baremetal-coco: Add DCAP deployment for TDX

Let's set up a per-cluster PCCS, meaning:
* a single service that chaces all the DCAP collateral requests
* "indirect" DCAP registration, which stores the encrypted platform keys
  on PCCS
* PCCS database on a dedicated node, where the admin knows what to back
  up (if needed)

This is required in order to have attestation working for TDX, be it
using DCAP directly or ITA.

Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>
Signed-off-by: Mikko Ylinen <mikko.ylinen@intel.com>

fidencio committed:
baremetal-coco: Remove intel-dpo on uninstall

We just missed doing this when we added the deployment of the intel
device plugins operator.

Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>

fidencio commented:
Please, pretty please, let's have it properly tested before merged. :-)

bpradipt commented on scripts/install-helpers/baremetal-coco/intel-dcap/pccs.yaml.in:
```diff
@@ -0,0 +1,123 @@
+        kubernetes.io/hostname: ${PCCS_NODE}
+      initContainers:
+       - name: init-seclabel
+         image: quay.io/fidencio/init-seclabel:latest
```
Note-to-self :-) 
After testing and before merging, we should put the images from personal registry to https://quay.io/openshift_sandboxed_containers/

bpradipt replied:
quay.io/openshift_sandboxed_containers/init-seclabel:latest

mythi commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -473,6 +589,30 @@ function verify_params() {
+    echo "Intel DCAP | starting the uninstall"
+
+    pushd intel-dcap || return 1
+    oc apply -f qgs.yaml || return 1
+    oc apply -f registration-ds.yaml || return 1
+    oc apply -f pccs.yaml || return 1
+    oc apply -f ns.yaml || return 1
```
`oc delete`?

fidencio replied:
Nice catch on my copy & paste. :-)

mythi commented on scripts/install-helpers/baremetal-coco/intel-dcap/pccs.yaml.in:
```diff
@@ -0,0 +1,123 @@
+      "AdminTokenHash" : "${PCCS_ADMIN_TOKEN_HASH}",
+      "CachingFillMode" : "LAZY",
+      "OPENSSL_FIPS_MODE" : false,
+      "LogLevel" : "debug",
```
The PCCS default is:
```suggestion
      "LogLevel" : "info",
```

mythi commented on scripts/install-helpers/baremetal-coco/intel-dcap/registration-ds.yaml.in:
```diff
@@ -0,0 +1,41 @@
+        securityContext:
+          readOnlyRootFilesystem: true
+          allowPrivilegeEscalation: true
+          privileged: true
```
we could explain the reason for privileged here too:
```suggestion
          privileged: true # Required to access host efivarfs
```

a future improvement could be to move the privileged part to an `initContainer`

mythi commented on scripts/install-helpers/baremetal-coco/intel-dcap/pccs.yaml.in:
```diff
@@ -0,0 +1,123 @@
+          "define": {
+            "freezeTableName": true
+          },
+          "logging" : true,
```
PCCS default:
```suggestion
          "logging" : false,
```

fidencio replied:
I'd prefer keeping at least some log enabled, as otherwise we'll make our lives quite hard when it comes to debugging it.  I'll tone down the log level as you suggested from `debug` to `info`.

Thanks!

mythi reviewed (commented):
minor suggestions

fidencio commented:
@bpradipt, this one is finally ready to be reviewed (and hopefully, merged!)

fidencio committed:
baremetal-coco: Set https / no_proxy for the agent

This is needed as that's the only way for the agent running inside the
guest to know about proxies.

Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>

bpradipt commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -8,6 +8,19 @@ KERNEL_CONFIG_MC_FILE="./96-kata-kernel-config-mc.yaml"
+CLUSTER_NO_PROXY="$(oc get proxy/cluster -o json | jq .spec.noProxy)"
+
+[ "$CLUSTER_HTTPS_PROXY" = "null" ] && CLUSTER_HTTPS_PROXY=""
+[ "$CLUSTER_NO_PROXY" = "null" ] && CLUSTER_NO_PROXY=""
```
This should be set after we have checked for oc and jq.
I think, it's better to make jq also mandatory. So you can move 'check_command jq' after "check_command oc", rename set_aa_kbc_params_for_kata_agent to something like set_kernel_params_for_kata_agent and set the proxy.

mythi replied:
without dependencies to external tool you can `oc get proxy/cluster -o jsonpath={.spec.httpsProxy}`

mythi replied:
I can see now that `jsonpath` is widely used below so perhaps just make consistent?

fidencio replied:
I've reworked this bit, please, see the latest push.

bpradipt commented on scripts/install-helpers/baremetal-coco/intel-dcap/pccs.yaml.in:
```diff
@@ -0,0 +1,123 @@
+           privileged: true  # Required for chcon to work on host files
+      containers:
+        - name: pccs
+          image: quay.io/fidencio/dcap/pccs:1.21
```
quay.io/openshift_sandboxed_containers/dcap/pccs:1.21

bpradipt commented on scripts/install-helpers/baremetal-coco/intel-dcap/qgs.yaml:
```diff
@@ -0,0 +1,73 @@
+      dnsPolicy: ClusterFirstWithHostNet
+      containers:
+        - name: tdx-qgs
+          image: quay.io/fidencio/dcap/tdx-qgs:1.21
```
quay.io/openshift_sandboxed_containers/dcap/tdx-qgs:1.21

bpradipt commented on scripts/install-helpers/baremetal-coco/intel-dcap/registration-ds.yaml.in:
```diff
@@ -0,0 +1,47 @@
+          - name: efivars
+            mountPath: /sys/firmware/efi/efivars
+        workingDir: "/opt/intel/sgx-pck-id-retrieval-tool/"
+        image: quay.io/fidencio/dcap/intel-dcap-registration-flow:devel
```
quay.io/openshift_sandboxed_containers/dcap/dcap-registration-flow:latest

mythi commented on scripts/install-helpers/baremetal-coco/intel-dcap/qgs.yaml:
```diff
@@ -0,0 +1,73 @@
+      annotations:
+        sgx.intel.com/quote-provider: tdx-qgs
+    spec:
+      hostNetwork: true
```
is this a leftover from when we had PCCS serving localhost? I think this should be dropped.

fidencio replied:
> is this a leftover from when we had PCCS serving localhost? I think this should be dropped.

This should **NOT** be dropped, otherwise we start having issues to start the pod due to SELInux shenanigans.

mythi replied:
I fail to understand how this relates to SELinux. something to do with the `vsock` setup?

fidencio replied:
I'm marking this one as resolved, as it doesn't make much sense spending time to improve this approach, as we'll be using the DCAP packages from Virt SIG soon enough.

fidencio committed:
baremetal-coco: Make jq required

As we're relying on `jq` for a few cases, let's just make sure this is
required by moving its check to right after the `oc ` check.

Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>

fidencio committed:
baremetal-coco: set_{aa_kbc,kernel}_params_for_kata_agent

Rename the set_aa_kbc_params_for_kata_agent to
set_kernel_params_for_kata_agent, as later in this series we'll also use
this function to set the agent.https_proxy and agent.no_proxy
parameters, which are needed in order to properly pull images inside the
guest on a cluster running behind proxies.

Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>

bpradipt commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -8,6 +8,16 @@ KERNEL_CONFIG_MC_FILE="./96-kata-kernel-config-mc.yaml"
 SKIP_NFD="${SKIP_NFD:-false}"
 TRUSTEE_URL="${TRUSTEE_URL:-"http://kbs-service.trustee-operator-system:8080"}"
 CMD_TIMEOUT="${CMD_TIMEOUT:-900}"
+CLUSTER_HTTPS_PROXY="$(oc get proxy/cluster -o jsonpath={.spec.httpsProxy})"
```
@fidencio we are still relying on `oc` and assignment should be after checking of `oc` command.
Also since we don't expect the admin to set this, I think better to have this assignment closer to the usage ie in `set_kernel_params_for_kata_agent`..

fidencio replied:
Fixed, please, check the latest version.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -8,6 +8,15 @@ KERNEL_CONFIG_MC_FILE="./96-kata-kernel-config-mc.yaml"
 SKIP_NFD="${SKIP_NFD:-false}"
 TRUSTEE_URL="${TRUSTEE_URL:-"http://kbs-service.trustee-operator-system:8080"}"
 CMD_TIMEOUT="${CMD_TIMEOUT:-900}"
+CLUSTER_HTTPS_PROXY="$(oc get proxy/cluster -o jsonpath={.spec.httpsProxy})"
```
This should still only be set after `oc` is known to be available.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -690,6 +831,7 @@ if [ "$SKIP_NFD" = false ]; then
     tdx)
         create_intel_node_feature_rules || exit 1
         deploy_intel_device_plugins || exit 1
+        deploy_intel_dcap || exit 1
```
IMHO it would be clearer to use the _install/uninstall_ wording instead of _deploy/uninstall_ but this can be fixed later as this _nit_ predates the current PR.

fidencio replied:
I won't change this for now, as the `deploy` pattern was there before my patches, and I'm just following that.

Later on we can decide to have a better naming, and change it all over the place, at once.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -9,6 +9,7 @@ SKIP_NFD="${SKIP_NFD:-false}"
 TRUSTEE_URL="${TRUSTEE_URL:-"http://kbs-service.trustee-operator-system:8080"}"
 CMD_TIMEOUT="${CMD_TIMEOUT:-900}"
 CLUSTER_HTTPS_PROXY="$(oc get proxy/cluster -o jsonpath={.spec.httpsProxy})"
+CLUSTER_NO_PROXY="$(oc get proxy/cluster -o jsonpath={.spec.noProxy})"
```
Same remark about `oc` being available.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -369,6 +371,11 @@ function set_kernel_params_for_kata_agent() {
+        kata_override="[hypervisor.qemu]
+kernel_params=\"agent.aa_kbc_params=cc_kbc::$trustee_url agent.https_proxy=$cluster_https_proxy agent.no_proxy=$cluster_no_proxy\""
+    fi
+
```
What about concatenating `"agent.https_proxy=..."` to the existing `kernel_params` instead of duplicating the code ? And do the same for `agent.no_proxy`.

fidencio replied:
Done, please, take a look.

mythi commented on scripts/install-helpers/baremetal-coco/intel-dcap/pccs.yaml.in:
```diff
@@ -0,0 +1,123 @@
+           privileged: true  # Required for chcon to work on host files
+      containers:
+        - name: pccs
+          image: quay.io/openshift_sandboxed_containers/dcap/pccs:1.21
```
nit: could also use pinned SHAs?

fidencio replied:
Let's skip it for now as we'll soon enough **NOT** use those containers and rely on the packages being part of Virt SIG.

There's no reason to over complicate this approach knowing it'll be replaced as soon as possible.

lmilleri commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -247,6 +278,40 @@ function deploy_intel_device_plugins() {
+    oc apply -f ns.yaml || return 1
+
+    oc project intel-dcap
+    oc adm policy add-scc-to-user privileged -z default
```
does the dcap need to run in a privileged mode?

fidencio replied:
Yes, it does, unfortunately.

For this, I think we can just leave it as it is for now, as soon enough we should not need to run DCAP / PCCS / QGS from inside a container, with everything being packaged as part of the virt sig.

fidencio commented:
I started testing this version of the PR, as it is ... let's see how it goes ...

gkurz reviewed (approved):
/lgtm

Thanks @fidencio !

> I started testing this version of the PR, as it is ... let's see how it goes ...

This PR could potentially skip CI since nothing calls this code but I have no control on that AFAIK. I'll thus start CI when you are done with your testing.

bpradipt reviewed (approved):
/lgtm

fidencio commented:
> I started testing this version of the PR, as it is ... let's see how it goes ...

Finished the tests on my side, on an environment that's not behind proxies and it works as expected.
I'd consider this good enough to be merged, and any other adjustment can come later on.
